### PR TITLE
8290368: Introduce LDAP and RMI protocol-specific object factory filters to JNDI implementation

### PR DESCRIPTION
**Pull request by AlekseiEfimov**
### Summary of the change
This change introduces new system and security properties for specifying factory filters for the JNDI/LDAP and the JNDI/RMI JDK provider implementations. 

These new properties allow more granular control over the set of object factories allowed to reconstruct Java objects from LDAP and RMI contexts.

The new factory filters are supplementing the existing `jdk.jndi.object.factoriesFilter` global factories filter to determine if a specific object factory is permitted to instantiate objects for the given protocol.

Links:

- [CSR with details](https://bugs.openjdk.org/browse/JDK-8291556)
- [JBS issue](https://bugs.openjdk.org/browse/JDK-8290368)

### List of code changes

- Implementation for two new system and security properties have been added to the `com.sun.naming.internal.ObjectFactoriesFilter` class
- `java.security` and `module-info.java` files have been updated with a documentation for the new properties
- To keep API of `javax.naming.spi.NamingManager` and `javax.naming.spi.DirectoryManager` classes unmodified a new internal `com.sun.naming.internal.NamingManagerHelper` class has been introduced. All  `getObjectInstance` calls have been updated to use the new helper class.

#### NamingManagerHelper changes
Changes performed to construct the `NamingManagerHelper` class:
- `DirectoryManager.getObjectInstance` -> `NamingManagerHelper.getDirObjectInstance`. Dependant methods were also moved to the `NamingManagerHelper` class
- `NamingManager.getObjectInstance` -> `NamingManagerHelper.getObjectInstance`. Methods responsible for setting/getting object factory builder were moved to the `NamingManagerHelper` class too.

### Test changes
New tests have been added for checking that new factory filters can be used to restrict reconstruction of Java objects from LDAP and RMI contexts:
- LDAP protocol specific test: test/jdk/com/sun/jndi/ldap/objects/factory/LdapFactoriesFilterTest.java
- RMI protocol specific test: test/jdk/com/sun/jndi/rmi/registry/objects/RmiFactoriesFilterTest.java
Existing `test/jdk/javax/naming/module/RunBasic.java` test has been updated to allow test-specific factories filter used to reconstruct objects from the test LDAP server. 

### Testing
tier1-tier3 and JNDI regression/JCK tests not showing any failures related to this change.
No failures observed for the modified regression tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8290368](https://bugs.openjdk.org/browse/JDK-8290368): Introduce LDAP and RMI protocol-specific object factory filters to JNDI implementation
 * [JDK-8291556](https://bugs.openjdk.org/browse/JDK-8291556): Introduce LDAP and RMI protocol-specific object factory filters to JNDI implementation (**CSR**)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to [4449dda1](https://git.openjdk.org/jdk/pull/10578/files/4449dda181727185aa201de92387b8b6dc4a8dfc)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [528489b0](https://git.openjdk.org/jdk/pull/10578/files/528489b0f1aee1da1ad848b17ae7819953d7225a)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [4449dda1](https://git.openjdk.org/jdk/pull/10578/files/4449dda181727185aa201de92387b8b6dc4a8dfc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10578/head:pull/10578` \
`$ git checkout pull/10578`

Update a local copy of the PR: \
`$ git checkout pull/10578` \
`$ git pull https://git.openjdk.org/jdk pull/10578/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10578`

View PR using the GUI difftool: \
`$ git pr show -t 10578`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10578.diff">https://git.openjdk.org/jdk/pull/10578.diff</a>

</details>

### CODE REVIEW

**Review by dfuch — APPROVED** [review]

**Comment by dfuch** [comment]
I am approving on the condition to make sure run all JNDI tests (as well as tier1, tier2, tier3) are run before integrating.

**Comment by dfuch** [comment]
Also please get an approval from a security-libs maintainer for the changes to the security properties file before integrating.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (RogerRiggs) [inline src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:91]: This construct in which the supplied lambda fills in the serialClass is pretty obscure. 
Perhaps the variable name can be "serialClass" to match the only non-default method in ObjectInputFilter would give a better hint.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (RogerRiggs) [inline src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:92]: The handling of the selection of the filter could be more direct.
You can change the argument to checkInput to be ObjectInputFilter and pass the desired filter; LDAP_FILTER, RMI_FITLER, or GLOBAL_FILTER.
And make the check of the GLOBAL_FILTER conditional on it not having already been evaluated.
Then it will be the case that there must always be a specific filter.

The callers are all local to the class and would change to pass the desired filter.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (RogerRiggs) [inline src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:173]: Where does the IllegalArgumentException from ObjectInputFilter.create get handled or reported?
If the property value is illformed, the error should be enough to diagnose and correct the property.

File: test/jdk/com/sun/jndi/ldap/objects/factory/LdapFactoriesFilterTest.java
Comment (RogerRiggs) [inline test/jdk/com/sun/jndi/ldap/objects/factory/LdapFactoriesFilterTest.java:46]: Are any of these filter property cases, malformed and would produce an error from ObjectInputFilter.create?

**Review by RogerRiggs — COMMENTED** [review]
Nice work, but a few comments.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java]: Thanks - `serialClass` name reads better.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java]: Thank you - refactored as suggested.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:174]: That is a good point - the current state of reporting for a malformed filter pattern can be improved:
First filter check with `check*Filter` throws `java.lang.ExceptionInInitializerError` with a cause set to `java.lang.IllegalArgumentException` with filter pattern error. But subsequent checks throw `java.lang.NoClassDefFoundError: Could not initialize class com.sun.naming.internal.ObjectFactoriesFilter`.

To address that one interface and two new records have been added to represent a factory filter state:
 - `ConfiguredFilter` - interface for representing a filter created with `ObjectInputFilter.create` from a pattern.
 - `ValidFilter implements ConfiguredFilter` - stores `ObjectInputFilter` constructed from a valid filter pattern.
 - `InvalidFilter implements ConfiguredFilter` - stores `IllegalArgumentException` generated by parsing of an invalid filter pattern. The stored `IAE` is used to report malformed filter pattern each time specific filter is consulted.

File: test/jdk/com/sun/jndi/ldap/objects/factory/LdapFactoriesFilterTest.java
Comment (AlekseiEfimov) [reply test/jdk/com/sun/jndi/ldap/objects/factory/LdapFactoriesFilterTest.java]: There were no such cases.
Now, RMI and LDAP tests have been modified to run with malformed filter patterns which produce an error from `ObjectInputFilter.create`. 
Also, new cases were added to check that illegal patterns in one of RMI or LDAP filters not preventing other from being used.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (dfuch) [inline src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:99]: If I'm not mistaken there's  no point in checking the specific filter if the global filter state is REJECTED. So instead of switching on the specificResult below, maybe you should change the logic to switch on the globalResult instead and only call the specific filter if needed?

**Review by dfuch — CHANGES_REQUESTED** [review]

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:99]: > If I'm not mistaken there's no point in checking the specific filter if the global filter state is REJECTED. So instead of switching on the specificResult below, maybe you should change the logic to switch on the globalResult instead and only call the specific filter if needed?

Yes - there is no point, and that will reduce number of `checkInput` calls on a specific filter, if it is not the global one. That's how it will look like:
```
    private static boolean checkInput(ConfiguredFilter filter, FactoryInfo serialClass) {
        var globalFilter = GLOBAL_FILTER.filter();
        var specificFilter = filter.filter();
        Status globalResult = globalFilter.checkInput(serialClass);

        // Check if a specific filter is the global one
        if (filter == GLOBAL_FILTER) {
            return globalResult == Status.ALLOWED;
        }
        return switch (globalResult) {
            case ALLOWED -> specificFilter.checkInput(serialClass) != Status.REJECTED;
            case REJECTED -> false;
            case UNDECIDED -> specificFilter.checkInput(serialClass) == Status.ALLOWED;
        };
    }
```

The `if (filter == GLOBAL_FILTER) {` check can be also removed but without it the `checkInput` will be called twice on global filter for the case where `specific == global`, ie call from the `checkGlobalFilter`.

Comment (dfuch) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:99]: The code above LGTM. An alternative could be:

```
    private static boolean checkInput(ConfiguredFilter filter, FactoryInfo serialClass) {
        var globalFilter = GLOBAL_FILTER.filter();
        var specificFilter = filter.filter();
        Status globalResult = globalFilter.checkInput(serialClass);

        return switch (globalResult) {
            case ALLOWED -> filter == GLOBAL || specificFilter.checkInput(serialClass) != Status.REJECTED;
            case REJECTED -> false;
            case UNDECIDED -> specificFilter.checkInput(serialClass) == Status.ALLOWED;
        };
    }
```

but I believe your proposed code reads better.

Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java]: Thanks - pushed as 528489b

**Review by dfuch — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]
LGTM

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (jaikiran) [inline src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:59]: I think the `serialClass` param should be renamed to `factoryClass` or something like that, since I think the `serialClass` reference comes from serialization/deserialization usage.

File: src/java.base/share/conf/security/java.security
Comment (jaikiran) [inline src/java.base/share/conf/security/java.security:1408]: Hello Aleksei, as far as I can see the `jdk.serialFilter` allows for additional details like `limits` (example: `maxdepth=value`). Should we note here that this JNDI specific property will ignore those configurations?

File: src/java.base/share/conf/security/java.security
Comment (jaikiran) [inline src/java.base/share/conf/security/java.security:1423]: Should this instead say:
> The default pattern value allows any object factory class belonging to the `com.sun.jndi.ldap` and its sub-packages of the `java.naming` module to be specified by the reference instance...?

File: src/java.naming/share/classes/javax/naming/spi/NamingManager.java
Comment (jaikiran) [inline src/java.naming/share/classes/javax/naming/spi/NamingManager.java:109]: Perhaps remove the `synchronized` from this method and the getter, now that the synchronization has moved to the `NamingManagerHelper` class? I don't see any other instance synchronization usages in this class, so removing the `synchronized` should be OK.

File: src/java.base/share/conf/security/java.security
Comment (jaikiran) [inline src/java.base/share/conf/security/java.security:1410]: I believe this should be `its` instead of `it's`

File: src/java.naming/share/classes/com/sun/naming/internal/NamingManagerHelper.java
Comment (jaikiran) [inline src/java.naming/share/classes/com/sun/naming/internal/NamingManagerHelper.java:81]: I think this "Will reach this point ..." part should be updated to note that the code will additionally reach here if the factory class is disallowed by the filters.

Similar change would be needed for the comment in `getDirObjectInstance` method.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:59]: The name comes from `ObjectInputFilter.FilterInfo` - it's been renamed from `factoryClass`to make it clear that the supplied lambda fills-in the non-default `ObjectInputFilter.FilterInfo.serialClass()` method.

File: src/java.naming/share/classes/javax/naming/spi/NamingManager.java
Comment (AlekseiEfimov) [reply src/java.naming/share/classes/javax/naming/spi/NamingManager.java:109]: Agreed - I think it is safe to remove `synchronized`. I doubt that there is a code that relies on `synchronized (javax.naming.spi.NamingManager.class)` to prevent other code setting the factory builder (also it can be set only once), therefore it should be ok to remove `synchronized`.

File: src/java.base/share/conf/security/java.security
Comment (AlekseiEfimov) [reply src/java.base/share/conf/security/java.security:1423]: I would prefer to keep a filter default value explanation sentences simpler since exact default values are listed just one line after.

File: src/java.base/share/conf/security/java.security
Comment (dfuch) [inline src/java.base/share/conf/security/java.security:1388]: It appears that for those protocol for which there is no specific filter, a factory class will be accepted only if the global filter returns ALLOWED - which contradicts what is written here (where it says that the class is allowed if it's not REJECTED). Is this something that has changed with this fix - or was the documentation wrong before?

Comment (AlekseiEfimov) [reply src/java.base/share/conf/security/java.security:1388]: Very good catch Daniel! It is with this fix and I believe the code needs to be change to match what is written for the global filter here:
What we've had before in `checkInput`:
```
    private static boolean checkInput(FactoryInfo factoryInfo) {
        Status result = GLOBAL.checkInput(factoryInfo);
        return result != Status.REJECTED;
```
What we have now:
```
        if (filter == GLOBAL_FILTER) {
            return globalResult == Status.ALLOWED;
        }
```

I think it needs to be changed to (to match the description for global filter):
```
        if (filter == GLOBAL_FILTER) {
            return globalResult != Status.REJECTED;
        }
```

Comment (RogerRiggs) [reply src/java.base/share/conf/security/java.security:1388]: In the general composition of filters, it is preferable that UNDECIDED is treated as REJECTED.
That keeps unintentional holes in a filter from being permissive.

File: src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java
Comment (jaikiran) [reply src/java.naming/share/classes/com/sun/naming/internal/ObjectFactoriesFilter.java:59]: Thank you for that clarification, Aleksei. This sounds fine to me then.

File: src/java.base/share/conf/security/java.security
Comment (jaikiran) [reply src/java.base/share/conf/security/java.security:1425]: What you have here is OK with me then.

File: src/java.base/share/conf/security/java.security
Comment (AlekseiEfimov) [reply src/java.base/share/conf/security/java.security:1388]: > In the general composition of filters, it is preferable that UNDECIDED is treated as REJECTED.
 > That keeps unintentional holes in a filter from being permissive.

That is a good point Roger. The "java.security" file was updated (4449dda) to match the `ObjectFactoriesFilter` implementation, ie the global filter treats UNDECIDED as REJECTED. Also, the CSR has been updated to highlight that.

**Review by jaikiran — APPROVED** [review]
The latest state of this PR in `4449dda` looks good to me. Thank you Aleksei.

**Review by dfuch — APPROVED** [review]

**Pull request closed without merging**
